Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift

```diff
@@ -76,23 +115,67 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
 private extension NoMagicNumbersRule {
     final class Visitor: ViolationsSyntaxVisitor {
         private let testParentClasses: Set<String>
+        private var testClasses: Set<String> = []
+        private var nonTestClasses: Set<String> = []
+        private var possibleViolations: [String: [ReasonedRuleViolation]] = [:]
 
         init(viewMode: SyntaxTreeViewMode, testParentClasses: Set<String>) {
             self.testParentClasses = testParentClasses
             super.init(viewMode: viewMode)
         }
 
+        override func visitPost(_ node: ClassDeclSyntax) {
+            let className = node.identifier.text
+            if node.isXCTestCase(testParentClasses) {
+                testClasses.insert(className)
+                removeViolations(forClassName: className)
+            } else {
+                nonTestClasses.insert(className)
+            }
+        }
+
         override func visitPost(_ node: FloatLiteralExprSyntax) {
-            if node.isMemberOfATestClass(testParentClasses) == false, node.floatingDigits.isMagicNumber {
-                violations.append(node.floatingDigits.positionAfterSkippingLeadingTrivia)
+            guard node.floatingDigits.isMagicNumber else {
+                return
             }
+            let violation = node.floatingDigits.positionAfterSkippingLeadingTrivia
+            process(violation: violation, forNode: node)
         }
 
         override func visitPost(_ node: IntegerLiteralExprSyntax) {
-            if node.isMemberOfATestClass(testParentClasses) == false, node.digits.isMagicNumber {
-                violations.append(node.digits.positionAfterSkippingLeadingTrivia)
+            guard node.digits.isMagicNumber else {
+                return
+            }
+            let violation = node.digits.positionAfterSkippingLeadingTrivia
+            process(violation: violation, forNode: node)
```

Comment:
`process` is not very descriptive. The position can be gathered from the node like `node.positionAfterSkippingLeadingTrivia`:
```suggestion
            collectViolations(forNode: node)
```